axon/threads.go

```diff
@@ -195,30 +195,25 @@ func (nt *NetworkBase) LayerFun(fun func(ly AxonLayer), funame string, thread, w
 // using Neurons threading (go routines) if thread is true and NThreads > 1.
 // if wait is true, then it waits until all procs have completed.
 func (nt *NetworkBase) NeuronFun(fun func(ly AxonLayer, ni int, nrn *Neuron), funame string, thread, wait bool) {
-	if thread { // don't bother if not significant to thread
-		nt.FunTimerStart(funame)
-	}
 	if !thread || nt.NThreads <= 1 {
-		for ni := range nt.Neurons {
-			nrn := &nt.Neurons[ni]
-			ly := nt.Layers[nrn.LayIdx].(AxonLayer)
-			fun(ly, ni-ly.NeurStartIdx(), nrn)
-		}
-		if thread {
-			nt.FunTimerStop(funame)
+		for _, layer := range nt.Layers {
+			lyr := layer.(AxonLayer)
+			lyrNeurons := lyr.AsAxon().Neurons
+			for nrnIdx := range lyrNeurons {
+				nrn := &lyr.AsAxon().Neurons[nrnIdx]
```

Comment:
why not: `nrn := lyrNeurons[nrnIdx]` ?

Reply:
Also did you specifically compare the case of making a local var of Neurons vs. a local var of lyr.AsAxon() and then doing the field access of Neurons through that?  `lyrNeurons := lyr.AsAxon().Neurons` makes a copy of the slice (fortunately not the *contents* of the slice!) which might be more expensive than doing the field access?  probably it doesn't make a diff I guess.

Reply:
`nrn := lyrNeurons[nrnIdx]` is the way to go, you're right. It doesn't make a diff in terms of runtime, I assume the compiler had already optimized it anyway (common subexpression elimination).

I tried your 2nd suggestion as well, but it doesn't make a difference. Based on my reading of [this post](https://research.swtch.com/godata), it sounds like copying a slice is dirt cheap as it just results in 3 new 64bit variables (ptr, len, capacity) which are kept in the registers / stack.